File: thain-core/src/main/java/com/xiaomi/thain/core/scheduler/job/FlowJob.java
/*
 * Copyright (c) 2019, Xiaomi, Inc.  All rights reserved.
 * This source code is licensed under the Apache License Version 2.0, which
 * can be found in the LICENSE file in the root directory of this source tree.
 */

package com.xiaomi.thain.core.scheduler.job;

import com.xiaomi.thain.common.constant.FlowExecutionStatus;
import com.xiaomi.thain.common.exception.ThainCreateFlowExecutionException;
import com.xiaomi.thain.common.exception.ThainRuntimeException;
import com.xiaomi.thain.common.model.dp.AddFlowExecutionDp;
import com.xiaomi.thain.core.constant.FlowExecutionTriggerType;
import com.xiaomi.thain.core.process.ProcessEngine;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.quartz.Job;
import org.quartz.JobExecutionContext;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import static com.xiaomi.thain.common.utils.HostUtils.getHostInfo;

/**
 * @author liangyongrui
 */
@Slf4j
public class FlowJob implements Job {

    @NonNull
    private final ProcessEngine processEngine;

    private static final Map<String, FlowJob> FLOW_JOB_MAP = new ConcurrentHashMap<>();

    private FlowJob(@NonNull ProcessEngine processEngine) {
        this.processEngine = processEngine;
    }

    public static FlowJob getInstance(@NonNull ProcessEngine processEngine) {
        return FLOW_JOB_MAP.computeIfAbsent(processEngine.processEngineId,
                t -> new FlowJob(processEngine));
    }

    @Override
    public void execute(@NonNull JobExecutionContext context) {
        try {
            long flowId = context.getJobDetail().getJobDataMap().getLong("flowId");
            val addFlowExecutionDp = AddFlowExecutionDp.builder()
                    .flowId(flowId)
                    .hostInfo(getHostInfo())
                    .status(FlowExecutionStatus.WAITING.code)
                    .triggerType(FlowExecutionTriggerType.AUTOMATIC.code)
                    .build();
            processEngine.processEngineStorage.flowExecutionDao.addFlowExecution(addFlowExecutionDp);
            if (addFlowExecutionDp.id == null) {
                throw new ThainCreateFlowExecutionException();
            }

            val flowExecutionDr = processEngine.processEngineStorage.flowExecutionDao
                    .getFlowExecution(addFlowExecutionDp.id).orElseThrow(ThainRuntimeException::new);
            processEngine.processEngineStorage.flowExecutionWaitingQueue.put(flowExecutionDr);
        } catch (Exception e) {
            log.error("Failed to add queue：", e);
        }
    }

}

File: thain-core/src/main/kotlin/com/xiaomi/thain/core/scheduler/job/FlowJob.kt
/*
 * Copyright (c) 2019, Xiaomi, Inc.  All rights reserved.
 * This source code is licensed under the Apache License Version 2.0, which
 * can be found in the LICENSE file in the root directory of this source tree.
 */
package com.xiaomi.thain.core.scheduler.job

import com.xiaomi.thain.common.constant.FlowExecutionStatus
import com.xiaomi.thain.common.exception.ThainCreateFlowExecutionException
import com.xiaomi.thain.common.exception.ThainRuntimeException
import com.xiaomi.thain.common.model.dp.AddFlowExecutionDp
import com.xiaomi.thain.common.utils.HostUtils
import com.xiaomi.thain.core.constant.FlowExecutionTriggerType
import com.xiaomi.thain.core.process.ProcessEngine
import org.quartz.Job
import org.quartz.JobExecutionContext
import org.slf4j.LoggerFactory
import java.util.concurrent.ConcurrentHashMap

/**
 * @author liangyongrui
 */
class FlowJob private constructor(private val processEngine: ProcessEngine) : Job {
    private val log = LoggerFactory.getLogger(this.javaClass)!!

    override fun execute(context: JobExecutionContext) {
        try {
            val flowId = context.jobDetail.jobDataMap.getLong("flowId")
            val addFlowExecutionDp = AddFlowExecutionDp(
                    flowId = flowId,
                    hostInfo = HostUtils.hostInfo,
                    status = FlowExecutionStatus.WAITING.code,
                    triggerType = FlowExecutionTriggerType.AUTOMATIC.code,
                    variables = "{}")
            processEngine.processEngineStorage.flowExecutionDao.addFlowExecution(addFlowExecutionDp)
            if (addFlowExecutionDp.id == null) {
                throw ThainCreateFlowExecutionException()
            }
            val flowExecutionDr = processEngine.processEngineStorage.flowExecutionDao
                    .getFlowExecution(addFlowExecutionDp.id!!) ?: throw ThainRuntimeException()
            processEngine.processEngineStorage.flowExecutionWaitingQueue.put(flowExecutionDr)
            log.debug("flow {} add queue，There are currently {} flows in the queue",
                    flowId, processEngine.processEngineStorage.flowExecutionWaitingQueue.size)
        } catch (e: Exception) {
            log.error("Failed to add queue：", e)
        }
    }

    companion object {
        private val FLOW_JOB_MAP: MutableMap<String, FlowJob> = ConcurrentHashMap()

        @JvmStatic
        fun getInstance(processEngine: ProcessEngine): FlowJob {
            return FLOW_JOB_MAP.computeIfAbsent(processEngine.processEngineId
            ) { FlowJob(processEngine) }
        }
    }

}
